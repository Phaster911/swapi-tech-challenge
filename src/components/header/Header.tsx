import { Search } from "react-feather";
import SearchInput from "../search-input/SearchInput";
import { FormEvent } from "react";
import styled from "styled-components";
import { COLORS, controlBorderRadius, QUERIES } from "../../style-constants";
import MaxWidthWrapper from "../shared/max-width-wrapper/MaxWidthWrapper";

const Wrapper = styled(MaxWidthWrapper)`
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr auto;
  margin: 0 auto;
  padding: 1rem 1rem;
  gap: 1rem;

  @media ${QUERIES.tabletAndUp} {
    grid-template-columns: minmax(max-content, 1fr) 1fr;
    grid-template-rows: 1fr;
  }
`;

const StyledH1 = styled.h1`
  text-align: center;

  @media ${QUERIES.tabletAndUp} {
    text-align: left;
  }
`;

const StyledForm = styled.form`
  display: flex;
  justify-content: center;
  gap: 1rem;

  @media ${QUERIES.tabletAndUp} {
    justify-content: flex-end;
  }
`;

const StyledSubmitButon = styled.button<{ disabled: boolean }>`
  border-radius: ${controlBorderRadius};
  padding: 1rem;
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  background-color: ${({ disabled }) =>
    disabled ? COLORS.mainBackground : COLORS.pageBackground};
  border: 2px solid ${COLORS.mainBackground};
`;

export type HeaderProps = {
  isPending: boolean;
  searchTerm: string;
  handleOnChangeSearch: (value: string) => void;
};

export default function Header({
  isPending,
  searchTerm,
  handleOnChangeSearch,
}: HeaderProps) {
  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
  };

  return (
    <header>
      <Wrapper>
        <StyledH1>Star Wars Characters</StyledH1>
        <StyledForm onSubmit={handleSubmit} role="search">
          <SearchInput
            disabled={isPending}
            value={searchTerm}
            onChangeHandler={handleOnChangeSearch}
          />
          <StyledSubmitButon
            type="submit"
            aria-label="Search for Star Wars characters by name"
            disabled={isPending}
            data-testid="search-form-submit-testid"
          >
            <Search
              aria-label="Search for Star Wars characters by name"
              color={COLORS.accent}
            />
          </StyledSubmitButon>
        </StyledForm>
      </Wrapper>
    </header>
  );
}
