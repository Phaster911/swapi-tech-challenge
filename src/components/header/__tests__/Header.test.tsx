import { render, screen } from "@testing-library/react";
import Header, { HeaderProps } from "../Header";

vi.mock("../../search-input/SearchInput", () => ({
  default: () => <div>SearchInput</div>,
}));

describe("<Header />", () => {
  let props: HeaderProps;
  let handleOnChangeSearchFn: ReturnType<typeof vi.fn>;
  beforeEach(() => {
    handleOnChangeSearchFn = vi.fn();
    props = {
      isPending: false,
      searchTerm: "",
      handleOnChangeSearch: handleOnChangeSearchFn,
    };
  });
  it("renders the component", () => {
    const { container } = render(<Header {...props} />);
    expect(container).toMatchSnapshot();
  });

  describe("when isPending is true", () => {
    it("should disable the form submit button", () => {
      const { container } = render(<Header {...props} isPending={true} />);
      expect(container).toMatchSnapshot();
      expect(screen.getByTestId("search-form-submit-testid")).toBeDisabled();
    });
  });
});
