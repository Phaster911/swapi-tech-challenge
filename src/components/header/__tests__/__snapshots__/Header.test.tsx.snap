// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > renders the component 1`] = `
.c0 {
  width: clamp(16rem,95vw,90rem);
}

.c1 {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr auto;
  margin: 0 auto;
  padding: 1rem 1rem;
  gap: 1rem;
}

.c2 {
  text-align: center;
}

.c3 {
  display: flex;
  justify-content: center;
  gap: 1rem;
}

.c4 {
  border-radius: 5px;
  padding: 1rem;
  cursor: pointer;
  background-color: white;
  border: 2px solid #e9ecee;
}

@media (min-width: 750px) {
  .c1 {
    grid-template-columns: minmax(max-content,1fr) 1fr;
    grid-template-rows: 1fr;
  }
}

@media (min-width: 750px) {
  .c2 {
    text-align: left;
  }
}

@media (min-width: 750px) {
  .c3 {
    justify-content: flex-end;
  }
}

<div>
  <header>
    <div
      class="c0 c1"
    >
      <h1
        class="c2"
      >
        Star Wars Characters
      </h1>
      <form
        class="c3"
        role="search"
      >
        <div>
          SearchInput
        </div>
        <button
          aria-label="Search for Star Wars characters by name"
          class="c4"
          data-testid="search-form-submit-testid"
          type="submit"
        >
          <svg
            aria-label="Search for Star Wars characters by name"
            fill="none"
            height="24"
            stroke="#5da647"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </button>
      </form>
    </div>
  </header>
</div>
`;

exports[`<Header /> > when isPending is true > should disable the form submit button 1`] = `
.c0 {
  width: clamp(16rem,95vw,90rem);
}

.c1 {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr auto;
  margin: 0 auto;
  padding: 1rem 1rem;
  gap: 1rem;
}

.c2 {
  text-align: center;
}

.c3 {
  display: flex;
  justify-content: center;
  gap: 1rem;
}

.c4 {
  border-radius: 5px;
  padding: 1rem;
  cursor: not-allowed;
  background-color: #e9ecee;
  border: 2px solid #e9ecee;
}

@media (min-width: 750px) {
  .c1 {
    grid-template-columns: minmax(max-content,1fr) 1fr;
    grid-template-rows: 1fr;
  }
}

@media (min-width: 750px) {
  .c2 {
    text-align: left;
  }
}

@media (min-width: 750px) {
  .c3 {
    justify-content: flex-end;
  }
}

<div>
  <header>
    <div
      class="c0 c1"
    >
      <h1
        class="c2"
      >
        Star Wars Characters
      </h1>
      <form
        class="c3"
        role="search"
      >
        <div>
          SearchInput
        </div>
        <button
          aria-label="Search for Star Wars characters by name"
          class="c4"
          data-testid="search-form-submit-testid"
          disabled=""
          type="submit"
        >
          <svg
            aria-label="Search for Star Wars characters by name"
            fill="none"
            height="24"
            stroke="#5da647"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </button>
      </form>
    </div>
  </header>
</div>
`;
